typescript/src/game_loop/game-rxjs.ts

```diff
@@ -1,115 +1,158 @@
+      next: (message) => queue.push(new MessageEnvelope(message, s2)),
+      complete: () => close(s1),
     })
+  );
+
+  subscriber.add(
+    getRxJSGameLoop(getTickRate()).subscribe((delta) => {
+      stats.addDelta(delta);
+
+      // 1. process messages
+      const messageEnvolopes = Array.from(queue);
+      queue.length = 0;
+      for (let i = 0; i < messageEnvolopes.length; i++) {
+        const m = messageEnvolopes[i];
+        world.processMessage(m.from, m.message);
+      }
+
+      // 2. update all positions
+      world.update(delta);
+
+      // 3. process collisions
+      world.collisions();
+
+      if (world.done) {
+        const gameResult = new GameResultsImpl(
+          stats,
+          world.getWinner(),
+          world.getLoser(),
+          false
+        );
+        subscriber.next(gameResult);
+        subscriber.complete();
+      }
+    })
+  );
 }
 
 export default function gameCreator(server: Server) {
-    server.on().pipe(
-        mergeMap(sockets => setupWithRxJS(sockets)),
-        tap(([s1, s2]) => {
-            s1.push(createMessage(MessageType.Play));
-            s2.push(createMessage(MessageType.Play));
-        }),
-        mergeMap(([s1, s2]) => {
-            GameStat.activeGames++;
-            return runRxJSLoop([s1, s2]).pipe(
-                tap({
-                    next: ([stats, winner, loser]) => {
-                        winner.push(createWinnerMessage(stats), () => winner.close());
-                        loser.push(createLoserMessage(), () => loser.close());
-                        GameStat.activeGames--;
-                    },
-                    error: (_) => {
-                        GameStat.activeGames--;
-                    }
-                }),
-                onErrorResumeNext(empty<GameResults>())
-            );
-        }),
-    ).subscribe(); // this sub is fine to live forever.  Once this is done, our
-    // server is done.
+  server
+    .on()
+    .pipe(mergeMap((sockets) => setupWithRxJS(sockets)))
+    .subscribe({
+      next: (results) => {
+        if (!results.disconnected) {
+          const { stats, winner, loser } = results;
+          winner.push(createWinnerMessage(stats), () => winner.close());
+          loser.push(createLoserMessage(), () => loser.close());
+        }
+        GameStat.activeGames--;
+      },
+    });
+}
+
+export function setupWithRxJS(
+  playerSockets: [RxSocket, RxSocket],
+  timeout: number = 30000
+): Observable<GameResults> {
+  return new Observable((subscriber) => {
+    const [p1, p2] = playerSockets;
+    p1.push(createReadyUpMessage());
+    p2.push(createReadyUpMessage());
+    let timeoutId = setTimeout(() => {
+      subscriber.error(new Error("Timeout"));
+    }, timeout);
+    subscriber.add(() => clearTimeout(timeoutId));
+
+    let readyCount = 0;
+    const checkReady = (msg: Message) => {
+      if (msg.type === MessageType.ReadyUp) {
+        readyCount++;
+        if (readyCount === 2) {
```

Comment:
Note that this assumes the clients are well-behaved. If one client sends two "ReadyUp" messages, this will break.